gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.gc;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.OptionalInt;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.spark.sql.SparkSession;
+import org.projectnessie.client.StreamingUtil;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.RefLogResponse;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.Tag;
+
+/**
+ * Encapsulates the logic to retrieve all content including their content keys over all commits in
+ * all named-references.
+ */
+public class GCImpl {
```

Comment:
One area that feels worth to be explored to save some work when traversing all references:
When you retrieve all references, you can get the common-ancestor with the default branch.
Assuming that the cut-off-time-interval for a reference is less-than-or-equal to the one of the default branch, you could actually stop traversing reference after processing the common ancestor commit from all references.

That's just an idea, not sure whether that works in practice - so nothing for this PR though - but probably worth a follow-up issue.